### PR TITLE
Change ban IP address field 

### PR DESCRIPTION
**Pull request by albertlast**
Man this is a mess.

i don't know if this pr is helpfull,
but it give a good overview where all problem raise up...

I implemented the pg and mysql support,
but i'm not able to test mysql.

One of the problems are, that pg take the ip address als "string" and mysql need it as integer,
reason for 'ip_address'.
Maybe would work for both string...

related to issue: https://github.com/SimpleMachines/SMF2.1/issues/2968

technical done,
~~upgrade is missing~~

### CODE REVIEW

**Comment by Oldiesmann** [comment]
I'm not sure how this is a good solution compared to how we're currently handling things. Also, this would require that we force people to have PHP compiled with bcmath (which is optional) and we'd need to do even more work in the upgrade to convert the existing IP bans, which could take quite some time on larger forums.

**Comment by albertlast** [comment]
The big pro for this kind of implementation is
"which could take quite some time on larger forums" here we can now use index.

The code is less complex.

The most code came from fallback for mysql and
the wildcard implementation, which not exist current.

There is no need to distinguish of ipv4 and 6.

Edit:
From my point of view the biggest disadvantages of the old solution ist,
that there no safe way if the ip address is are ipv4 or ipv6.
127|0|0|1|0|0|0|0 could be a ipv4 but 127:0:0:1:0:0:0:0 is a valid ipv6 too.

**Comment by jdarwood007** [comment]
The bcmath thing is important since its optional.

@Oldiesmann,  Upgrading shouldn't be too difficult.  We have the upgrader add the low and high columns and just call a massive update query.
Something like such:

```
UPDATE ban_items
SET ip_low = IF(ip_low5 > 0, CONCAT(ip_low1, ':', ip_low2, ':', ip_low3, ':', ip_low4, ':', ip_low5, ':', ip_low6, ':', ip_low7, ':', ip_low8), CONCAT(ip_low1, '.', ip_low2, '.', ip_low3, '.', ip_low4))
WHERE ip_low1 > 0
```

Adding the where into there would be ok so we can still do it in 1,000 item batches.

**Comment by albertlast** [comment]
well it's importent to know that the ip_low and ip_high are varbinary in mysql,
so a function update sql would be:

``` sql
UPDATE ban_items
SET ip_low = IF(ip_low5 > 0, 
                INET6_ATON(
                  CONCAT(ip_low1, ':', ip_low2, ':', ip_low3, ':', ip_low4, ':', ip_low5, ':', ip_low6, ':', ip_low7, ':', ip_low8)
                  )
                  , INET6_ATON(
                    CONCAT(ip_low1, '.', ip_low2, '.', ip_low3, '.', ip_low4)
                  )
               )
WHERE ip_low is null
```

But this only works on mysql version where the function INET6_ATON exists,
the version where you can find this is >5.6.2, mariadb >10.0.12 or in any mysql where this udf is exists:
https://bitbucket.org/skion/mysql-udf-ipv6

on every other version you need to load this data in php and use the fallback functions to convert this in right format.
Here a link to my test case for mysql: http://sqlfiddle.com/#!9/7e4cd/1/0

PG upgrade:

``` sql
UPDATE smf_ban_items
SET ip_low = (
    case when ip_low5 > 0 then CONCAT(ip_low1, ':', ip_low2, ':', ip_low3, ':', ip_low4, ':', ip_low5, ':', ip_low6, ':', ip_low7, ':', ip_low8) 
    else CONCAT(ip_low1, '.', ip_low2, '.', ip_low3, '.', ip_low4) 
    END
    )::inet
WHERE ip_low is null
```

```
```

**Comment by jdarwood007** [comment]
Is the IPv4 version also using INET6_ATON?  We could make it a CASE usage in MySQL to limit our need to specifics, but I see we would have to use ::net at the end, which MySQL doesn't do.

**Comment by albertlast** [comment]
This "::net" only works in postgresql.

For mysql it's essential that the INET6_ATON from mysql is used for ipv4 and ipv6 or
the fallback methode from Security.php inet_ptod.

I should mention that definition of "-" is changed in this pr.

It's look like in the current develop version is the - used to define a range of a block of the ip address.
In this pr is the - use to splitt the lower and the upper bound of the whole ip address.
current: 192.168.0-10.\* is internal 192.168.0.0 - 192.168.10.255
pr: 192.168.*-192.168.10.\* is internal 192.168.0.0 - 192.168.10.255

I already mention that i finde the current solution to complex internaly and for the user the benefit is smale or negative when he use ipv6 (with the short version of 0000:0000 -> ::)

**Comment by albertlast** [comment]
For me was on to many places a if for postgresql or not.
So i decided to add a new datatype "inet" to Subs-Db-\* in which the whole magic is done.

This pr use the function to write null values to the database,
to do so is the pr needed: https://github.com/SimpleMachines/SMF2.1/pull/3241

**Comment by Oldiesmann** [comment]
I still don't like that this uses the BCMath functions since that's an optional extension. Do we really need to use those instead of regular math functions?

**Comment by albertlast** [comment]
For me is important that we got the same result as the nativ function of mysql,
to keep flexible for the future.

But I'm not deep enough in php to get another way as this which I copy from stackoverflow.

The intention here to use bcmath I think is that ipv6 got a 128 bit integer.

**Comment by albertlast** [comment]
I don't believe it,
but it's look like the solution could be very easy for mysql.

See the last commit, would be good if some there with mysql and can test this.

Added {nullVar} as null string identifier.

Here the mysql upgrade command

``` sql
UPDATE ban_items
SET ip_low = case when ip_low8 > 0 THEN
    UNHEX(
        CONCAT(LPAD(hex(ip_low1),4,'0'),LPAD(hex(ip_low2),4,'0'),LPAD(hex(ip_low3),4,'0'),
        LPAD(hex(ip_low4),4,'0'),LPAD(hex(ip_low5),4,'0'),LPAD(hex(ip_low6),4,'0'),LPAD(hex(ip_low7),4,'0'),LPAD(hex(ip_low8),4,'0'))
    )
ELSE
    unhex(
        hex(
            INET_ATON(concat(ip_low1,'.',ip_low2,'.',ip_low3,'.',ip_low4))
        )
    )
END,
ip_high = case when ip_high8 > 0 THEN
    UNHEX(
        CONCAT(LPAD(hex(ip_high1),4,'0'),LPAD(hex(ip_high2),4,'0'),LPAD(hex(ip_high3),4,'0'),
        LPAD(hex(ip_high4),4,'0'),LPAD(hex(ip_high5),4,'0'),LPAD(hex(ip_high6),4,'0'),LPAD(hex(ip_high7),4,'0'),LPAD(hex(ip_high8),4,'0'))
    )
ELSE
    unhex(
        hex(
            INET_ATON(concat(ip_high1,'.',ip_high2,'.',ip_high3,'.',ip_high4))
        )
    )
END
```

Should again someone test with mysql.

**Comment by albertlast** [comment]
- Implemented db driven upgrade path
- found a bug in existing solution that the smallint on pg is to small for FFFF::

File: Sources/Subs.php [hunk 1]
```diff
@@ -3433,53 +3433,73 @@ function getAttachmentFilename($filename, $attachment_id, $dir = null, $new = fa
...
+			{
+				$ip_parts[1] .= $mode . $max;
+				$valid_high = inet_pton($ip_parts[1]);
+				$count++;
+				if ($count > 9) break;
+			}
+		}
+	    
+	    if($valid_high && $valid_low)
+	    {
```
Comment (live627) [inline Sources/Subs.php]: I see mixed spaces and tabs. Please use only tabs to indent the code.

File: other/upgrade_2-1_mysql.sql [hunk 2]
```diff
@@ -1908,4 +1880,28 @@ DROP INDEX idx_topic on {$db_prefix}messages;
...
+--- update ban ip with ipv6 support
+/******************************************************************************/
+
+ALTER TABLE {$db_prefix}ban_items ADD COLUMN ip_low varbinary(16);
+ALTER TABLE {$db_prefix}ban_items ADD COLUMN ip_high varbinary(16);
+
+UPDATE {$db_prefix}ban_items
+SET ip_low = 
+    UNHEX(
+        hex(
```
Comment (live627) [inline other/upgrade_2-1_mysql.sql:1895]: Why hex and then unhex? Am I missing something?

Comment (albertlast) [inline other/upgrade_2-1_mysql.sql:1895]: I don't know another way to convert this in varbinary.
the hex convert the int to hex,
the unhex convert this to varbinary.

**Comment by live627** [comment]
I really want to merge this in. Anything left that needs doing here?

**Comment by albertlast** [comment]
From my point of view I didn't see anything to do.

**Comment by live627** [comment]
I'm going to merge this in now... because I can, I suppose?

/me waits for the fallout

**Pull request merged**